sql/mysqld.cc

```diff
@@ -5397,21 +5397,30 @@ static int init_server_components()
     (void) mi_log(1);
 
 #if defined(HAVE_MLOCKALL) && defined(MCL_CURRENT) && !defined(EMBEDDED_LIBRARY)
-  if (locked_in_memory && !getuid())
+  if (locked_in_memory)
   {
-    if (setreuid((uid_t)-1, 0) == -1)
-    {                        // this should never happen
-      sql_perror("setreuid");
-      unireg_abort(1);
-    }
+    /* Try without priv user first, CAP_IPC_LOCK may be set */
     if (mlockall(MCL_CURRENT))
     {
-      if (global_system_variables.log_warnings)
-	sql_print_warning("Failed to lock memory. Errno: %d\n",errno);
-      locked_in_memory= 0;
+      int mlockall_err = errno;
+      /* ok, denied. Try to gain effective uid=0 to mlockall */
+      if (errno == EPERM && setreuid((uid_t)-1, 0) != -1)
+      {
+        if (mlockall(MCL_CURRENT)) {
+          if (global_system_variables.log_warnings)
+            sql_print_warning("Failed to lock memory as effective uid=0. Errno: %d\n",errno);
+          locked_in_memory= 0;
+        }
+        if (user_info)
+          set_user(mysqld_user, user_info);
+      }
+      else
+      {
+        if (global_system_variables.log_warnings)
+          sql_print_warning("Failed to lock memory as current user (didn't try as effective uid=0). Errno: %d\n", mlockall_err);
+        locked_in_memory= 0;
+      }
     }
-    if (user_info)
-      set_user(mysqld_user, user_info);
```

Comment:
if first mlockall() succeeds (or fails for reason other than EPERM), setreuid isn't called (the if takes a shortcut to fail because of order of the if),

This is the same as the existing implementation
https://github.com/MariaDB/server/blob/10.1/sql/mysqld.cc#L5399, set_user is only called if setreuid succeeds.

<pre>
Nov 13 10:42:55 spaceman mysqld[15250]: 2015-11-13 10:42:55 139814349588672 [Note] Plugin 'FEEDBACK' is disabled.
Nov 13 10:42:56 spaceman mysqld[15250]: 2015-11-13 10:42:56 139814349588672 [Warning] Failed to lock memory as current user (Errno: 12)
Nov 13 10:42:56 spaceman mysqld[15250]: 2015-11-13 10:42:56 139814349588672 [Warning] Failed setreuid to effective uid=0 while attempting to lock memory, (Errno: 12)
Nov 13 10:42:56 spaceman mysqld[15250]: 2015-11-13 10:42:56 139814349588672 [Note] Server socket created on IP: '::'.

ps -ef | grep mysqld
mysql    15250     1  0 10:42 ?        00:00:00 /usr/sbin/mysqld
<pre>

This was before I corrected the last != EPERM to == EPERM in the error reporting and hence the second message.

After compiling correction:


<pre>
2015-11-13 12:05:59 140169969162432 [Warning] Failed to lock memory as current user (Errno: 12)

ps -ef | grep mysqld
mysql    23079 23078  1 12:05 pts/2    00:00:00 sql/mysqld --skip-networking --datadir=/tmp/datadir --log-bin=/tmp/datadir/mysqlbin --socket /tmp/s.sock --lc-messages-dir=/home/dan/software_projects/mariadb-server/sql/share --verbose --user mysql --memlock
</pre>